File: mod/terraform-crud-api/resources/entity.lua
local resource_db = require('resource_db')

local function table_invert(t)
  local s={}
  for k,v in pairs(t) do
    s[v]=k
  end
  return s
end

local direction_to_string = table_invert(defines.direction)

local function get_direction(direction_name)
  local direction = defines.direction[direction_name]
  if direction == nil then
    error(string.format('Expected valid direction but got "%s"', direction_name))
  end
  return direction
end

local function entity_to_resource(e)
  return {
    unit_number = e.unit_number,
    surface = e.surface.name,
    name = e.name,
    position = e.position,
    direction = direction_to_string[e.direction],
    force = e.force.name
  }
end

return {
  read = function(query)
    local unit_number = query.unit_number
    local entity = resource_db.get('entity', unit_number)
    if entity == nil then
      return nil
    end
    if not entity.valid then
      resource_db.put('entity', unit_number, nil)
      return nil
    end
    return entity_to_resource(entity)
  end,

  create = function(config)
    local surface = game.surfaces[config.surface]
    if surface == nil then
      error(string.format('Could not find surface with id "%s"', config.surface))
    end
    local entity_creation_params = {
      name = config.name,
      position = config.position,
      direction = get_direction(config.direction),
      force = config.force,
      target = nil,
      source = nil,
      fast_replace = false,
      player = nil,
      spill = true,
      raise_built = true,
      create_build_effect_smoke = false,
      spawn_decorations = true,
      move_stuck_players = true,
      item = nil,
    }
    if config.entity_specific_parameters ~= nil then
      for k,v in pairs(config.entity_specific_parameters) do
        entity_creation_params[k] = v
      end
    end
    local e = surface.create_entity(entity_creation_params)
    if e == nil then
      error(string.format('Failed to create "%s"', config.name))
    end
    resource_db.put('entity', e.unit_number, e)
    return entity_to_resource(e)
  end,

  update = function(resource_id, update_config)
    local unit_number = tonumber(resource_id)
    local entity = resource_db.get('entity', unit_number)
    if entity == nil then
      return nil
    end
    if not entity.valid then
      resource_db.put('entity', unit_number, nil)
      return nil
    end
    if update_config.direction ~= nil then
      entity.direction = get_direction(update_config.direction)
    end
    if update_config.force ~= nil then
      entity.force = update_config.force
    end

    return entity_to_resource(entity)
  end,

  delete = function(resource_id)
    local unit_number = tonumber(resource_id)
    local entity = resource_db.get('entity', unit_number)
    if entity == nil then
      return {
        resource_exists = false
      }
    end
    if not entity.valid then
      resource_db.put('entity', unit_number, nil)
      return {
        resource_exists = false
      }
    end
    local destroyed = entity.destroy()
    if destroyed then
      resource_db.put('entity', unit_number, nil)
    end
    return {
      resource_exists = not destroyed
    }
  end,
}

File: provider/client/entity.go
package client

import "strconv"

type Position struct {
	X float32 `json:"x"`
	Y float32 `json:"y"`
}

type UnitNumber uint32

func (u UnitNumber) String() string {
	return strconv.FormatUint(uint64(u), 10)
}

func ParseUnitNumber(unitNumberStr string) (UnitNumber, error) {
	unitNumber, err := strconv.ParseUint(unitNumberStr, 10, 32)
	if err != nil {
		return 0, err
	}
	return UnitNumber(unitNumber), nil
}

type Entity struct {
	UnitNumber UnitNumber `json:"unit_number"`
	Surface    string     `json:"surface"`
	Name       string     `json:"name"`
	Position   Position   `json:"position"`
	Direction  Direction  `json:"direction"`
	Force      string     `json:"force"`
}

// Looks up an entity by its unit_number
func (client *FactorioClient) EntityGet(unitNumber UnitNumber) (*Entity, error) {
	var entity Entity
	pEntity := &entity
	return pEntity, client.Read(
		"entity",
		map[string]interface{}{"unit_number": unitNumber},
		&pEntity)
}

// Corresponding to https://lua-api.factorio.com/latest/LuaSurface.html#LuaSurface.create_entity
type EntityCreateOptions struct {
	Surface                  string                 `json:"surface"` // eg. "nauvis"
	Name                     string                 `json:"name"`
	Position                 Position               `json:"position"`
	Direction                Direction              `json:"direction"`
	Force                    string                 `json:"force"` // eg. "player", "enemy", "neutral"
	EntitySpecificParameters map[string]interface{} `json:"entity_specific_parameters"`

	// Unimplemented
	/*
		target;
		source;
		fast_replace;
		player;
		spill;
		raise_built;
		create_build_effect_smoke;
		spawn_decorations;
		move_stuck_players;
		item;
	*/
}

func (client *FactorioClient) EntityCreate(opts *EntityCreateOptions) (*Entity, error) {
	var result Entity
	return &result, client.Create("entity", opts, &result)
}

// All params are optional
type EntityUpdateOptions struct {
	Direction *Direction `json:"direction,omitempty"`
	Force     *string    `json:"force,omitempty"`
}

func (client *FactorioClient) EntityUpdate(unitNumber UnitNumber, opts *EntityUpdateOptions) (*Entity, error) {
	var result Entity
	return &result, client.Update(
		"entity",
		strconv.FormatUint(uint64(unitNumber), 10),
		opts,
		&result)
}

func (client *FactorioClient) EntityDelete(unitNumber UnitNumber) error {
	return client.Delete("entity", strconv.FormatUint(uint64(unitNumber), 10))
}
